### PR TITLE
TPC: Generic Check of Pads

### PR DESCRIPTION
**Pull request by HorstMa**
This PR is a generic check of the 72 TPC pads for an expected mean value, a comparison of the mean of one pad to the global mean and a check for empty bins inside the pads
This is the same as #1291, but with the changes implemented.

### CODE REVIEW

**Comment by Barthelemy** [comment]
I'll let @wiechula review.

File: Modules/TPC/include/TPC/CheckOfPads.h
Comment (wiechula) [inline Modules/TPC/include/TPC/CheckOfPads.h:51]: Naming convention
```suggestion
  std::vector<std::string> mSectorsNameEV;
  std::vector<Quality> mSectorsQualityEV;
  std::vector<std::string> mSectorsNameMean;
  std::vector<Quality> mSectorsQualityMean;
  std::vector<Quality> mSectorsQualityEmpty;
```

File: Modules/TPC/src/CheckOfPads.cxx
Comment (wiechula) [inline Modules/TPC/src/CheckOfPads.cxx:315]:
```suggestion
    auto* tcanv = (TCanvas*)mo->getObject();
    const int padsstart = 1;
    const int padsTotal = 72;
```

File: Modules/TPC/src/CheckOfPads.cxx
Comment (wiechula) [inline Modules/TPC/src/CheckOfPads.cxx:318]:
```suggestion
    const size_t end = moName.find("_2D");
    const auto histSubName = moName.substr(7, end - 7);
    const auto histNameS = fmt::format("h_{}_ROC", histSubName);
```

File: Modules/TPC/src/CheckOfPads.cxx
Comment (wiechula) [inline Modules/TPC/src/CheckOfPads.cxx:329]:
```suggestion
      const auto histName = fmt::format("{}_{:02d}", histNameS, tpads - 1);
      const auto h = (TH1F*)pad->GetListOfPrimitives()->FindObject(histName.data());
```

File: Modules/TPC/src/CheckOfPads.cxx
Comment (wiechula) [inline Modules/TPC/src/CheckOfPads.cxx:35]:
```suggestion
  if (const auto param = mCustomParameters.find("mediumQualityPercentageOfWorkingPads"); param != mCustomParameters.end()) {
```

File: Modules/TPC/src/CheckOfPads.cxx
Comment (wiechula) [inline Modules/TPC/src/CheckOfPads.cxx:41]:
```suggestion
  if (const 
  auto param = mCustomParameters.find("badQualityPercentageOfWorkingPads"); param != mCustomParameters.end()) {
```

File: Modules/TPC/src/CheckOfPads.cxx
Comment (wiechula) [inline Modules/TPC/src/CheckOfPads.cxx:35]: Here is a lot of copy'n'paste. And there is already a function for this: https://github.com/AliceO2Group/QualityControl/blob/352b520fbc9187ed7e857e263946d3c6b0580ce6/Modules/Common/include/Common/Utils.h
Used e.g. here: https://github.com/AliceO2Group/QualityControl/blob/master/Modules/TPC/src/PID.cxx#L46
Perhaps you can patch `Common/Utils.h` with
```diff
diff --git a/Modules/Common/include/Common/Utils.h b/Modules/Common/include/Common/Utils.h
index ad0127ca..15b14e8b 100644
--- a/Modules/Common/include/Common/Utils.h
+++ b/Modules/Common/include/Common/Utils.h
@@ -30,15 +30,14 @@ namespace o2::quality_control_modules::common
 /// \param name Name of the taskParameter
 /// \return taskParameter converted to bool, int, float, double or std::string depending on the template type
 template <typename T>
-T getFromConfig(const std::unordered_map<std::string, std::string>& params, const std::string_view name)
+T getFromConfig(const std::unordered_map<std::string, std::string>& params, const std::string_view name, T retVal = T{})
 {
   const auto last = params.end();
   const auto itParam = params.find(name.data());
-  T retVal{};
 
   if (itParam == last) {
     LOGP(warning, "missing parameter {}", name.data());
-    LOGP(warning, "Please add '{}': '<value>' to the 'taskParameters'.", name.data());
+    LOGP(warning, "Please add '{}': '<value>' to the 'taskParameters'. Using default value {}.", name.data(), retVal);
   } else {
     const auto& param = itParam->second;
     if constexpr (std::is_same<int, T>::value) {
```
and update all occurrances below with something like:
```c++
mMediumQualityLimit = common::getFromConfig<float>(mCustomParameters, "mediumQualityPercentageOfWorkingPads", 0.7);
```

File: Modules/TPC/src/CheckOfPads.cxx
Comment (wiechula) [inline Modules/TPC/src/CheckOfPads.cxx:135]:
```suggestion
    const auto moName = mo->getName();
```

File: Modules/TPC/src/CheckOfPads.cxx
Comment (wiechula) [inline Modules/TPC/src/CheckOfPads.cxx:138]:
```suggestion
      const size_t end = moName.find("_2D");
      const auto histSubName = moName.substr(7, end - 7);
```

File: Modules/TPC/src/CheckOfPads.cxx
Comment (wiechula) [inline Modules/TPC/src/CheckOfPads.cxx:206]: No need to loop twice
```suggestion
          for (int xBin = 1; xBin <= MaximumXBin; xBin++) {
            for (int yBin = 1; yBin <= MaximumYBin; yBin++) {
              const float binvalue = h->GetBinContent(xBin, yBin);
              if (binvalue != 0) {
                  padMean += binvalue;
                  padStdev += binvalue * binvalue;
                  padsCount++;
              }
            }
          }
        }
       if (padsCount>0) {
          padMean /= padsCount;
          padStdev = std::sqrt(std::abs(padStdev / padsCount - mean * mean));
        }
```

File: Modules/TPC/src/CheckOfPads.cxx
Comment (wiechula) [inline Modules/TPC/src/CheckOfPads.cxx:180]: Not needed
```suggestion
```

File: Modules/TPC/src/CheckOfPads.cxx
Comment (wiechula) [inline Modules/TPC/src/CheckOfPads.cxx:220]: Not needed, or to be modified above.
```suggestion
```

File: Modules/TPC/src/CheckOfPads.cxx
Comment (wiechula) [inline Modules/TPC/src/CheckOfPads.cxx:217]: totalPads can be `0` if title does not contain `IROC` or `OROC`. Either take care above and directly continue, or check for division by 0.

File: Modules/TPC/src/CheckOfPads.cxx
Comment (wiechula) [inline Modules/TPC/src/CheckOfPads.cxx:237]:
```suggestion
        if (mExpectedValueCheck) {
```

File: Modules/TPC/src/CheckOfPads.cxx
Comment (wiechula) [inline Modules/TPC/src/CheckOfPads.cxx:249]:
```suggestion
        if (mMeanCheck) {
```

File: Modules/TPC/src/CheckOfPads.cxx
Comment (wiechula) [inline Modules/TPC/src/CheckOfPads.cxx:273]:
```suggestion

        if (qualities.size() > 0) {
```

**Review by wiechula — CHANGES_REQUESTED** [review]

File: Modules/TPC/src/CheckOfPads.cxx
Comment (wiechula) [inline Modules/TPC/src/CheckOfPads.cxx:38]: Naming convention and logic
```suggestion
  std::string checkChoiceString = common::getFromConfig<std::string>(mCustomParameters, "CheckChoice", "Mean");
```
If the default value is `Mean`, the `mFoundCheck` will always be true. Is this the intended behaviour? If not, I think you should simple remove the `, "Mean"` part. Otherwise you can remove the lines 53-56 below, because this part of the code will never be executed...

Also `mFoundCheck` only seems to be used in this function, so no need for a member variable... Depending on the question above, it might even be eliminated completely.

File: Modules/TPC/src/CheckOfPads.cxx
Comment (wiechula) [inline Modules/TPC/src/CheckOfPads.cxx:57]:
```suggestion
```

File: Modules/TPC/src/CheckOfPads.cxx
Comment (wiechula) [inline Modules/TPC/src/CheckOfPads.cxx:56]: Why not
```suggestion
  bool foundCheck = false;
  if (checkChoiceString == CheckChoiceExpectedValue) {
    mExpectedValueCheck = true;
    foundCheck = true;
  }
  if (checkChoiceString == CheckChoiceMean) {
    mMeanCheck = true;
    foundCheck = true;
  }
  if (checkChoiceString == CheckChoiceEmpty) {
    mEmptyCheck = true;
    foundCheck = true;
  }
  
    if (!foundCheck) {
    mMeanCheck = true;
    ILOG(Warning, Support) << fmt::format("This Check requires a CheckChoice. There was no value or the given value is wrong or not readable. Chose between '{}'(Compare the pad mean to an expectedValue), '{}' (compare pad mean to global mean) and/or '{}' (check for a threshold of empty pads). As a default '{}' was selected.", CheckChoiceExpectedValue, CheckChoiceMean, CheckChoiceEmpty, CheckChoiceMean) << ENDM;
  }
```

and remove `mFoundCheck`.

**Review by wiechula — CHANGES_REQUESTED** [review]

File: Modules/TPC/src/CheckOfPads.cxx
Comment (wiechula) [reply Modules/TPC/src/CheckOfPads.cxx:38]: what about `const` and naming convention?

File: Modules/TPC/src/CheckOfPads.cxx
Comment (wiechula) [reply Modules/TPC/src/CheckOfPads.cxx]: what about the simplification using direct string comparison?

Comment (HorstMa) [reply Modules/TPC/src/CheckOfPads.cxx]: this would not work since we want to be able to run multiple checks at the same time (e.g. empty pad check and comparison against an expected value). So the checkChoiceString could look something like "ExpectedValue,Empty" so a direct string comparison would not work.

**Comment by Barthelemy** [comment]
we need a new approval

File: Modules/TPC/src/CheckOfPads.cxx
Comment (wiechula) [reply Modules/TPC/src/CheckOfPads.cxx]: Ahh, thanks for the explanation! That makes sense.

**Review by wiechula — APPROVED** [review]

**Review by Barthelemy — APPROVED** [review]

**Pull request merged**
